src/codegen/monomorphize.lisp

```diff
@@ -182,8 +183,14 @@ recompilation, and also maintains a stack of uncompiled candidates."
 
     ;; Fully propigating arguments could change the ordering of side effects
     ;; Exit early to avoid this
+    ;;
+    ;; Unless the function is known to have more arguments than are being applied
     (when (every #'propagatable-p new-args)
-      (return-from valid-candidate-p))
+      (let ((function (tc:lookup-function env name :no-error t)))
+        (unless function
+          (return-from valid-candidate-p))
```

Comment:
can you explicitly return NIL here and in the other cases ?

Reply:
Is that standard style in common lisp? I feel like it just add line noise

Reply:
I think it's like `when`/`unless` vs `cond`/`if`. If the return value matters (in this case it does), them being explicit is better than implicit. If the return value doesn't matter (a function just for effect), then eliding the return value is OK.

A non-expert Lisp programmer might also return `(return-from f)` as `f` returning `(values)`, which is not correct, but not an unreasonable assumption.